File: code/ex701.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */
class Solution {
    public TreeNode insertIntoBST(TreeNode root, int val) {
        if (root == null) {
            root = new TreeNode(val);
            return root;
        }
        
        TreeNode p = root;
        while (true) {
            if (val == p.val)    return root;
            else if (val > p.val) {
                if (p.right != null) {
                    p = p.right;
                    continue;
                } else {
                    p.right = new TreeNode(val);
                    return root;
                }
            }
            else {
                if (p.left != null) {
                    p = p.left;
                    continue;
                } else {
                    p.left = new TreeNode(val);
                    return root;
                }
            }
        }
    }
}

File: code/ex701.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    TreeNode* insertIntoBST(TreeNode* root, int val) {
        if (root == NULL)    return new TreeNode(val);
        
        TreeNode* p = root;
        while (1) {
            if (val < p->val) {
                if (p->left == NULL) {
                    p->left = new TreeNode(val);
                    return root;
                }
                p = p->left;
            } else {
                if (p->right == NULL) {
                    p->right = new TreeNode(val);
                    return root;
                }
                p = p->right;
            }
        }
        
        return root;
    }
};